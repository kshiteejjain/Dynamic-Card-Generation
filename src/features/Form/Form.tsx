import React from 'react';
import { useState } from "react";
import { useDispatch } from "react-redux";
import { useNavigate } from "react-router-dom";
import { addUser } from "./FormSlice";
import arrow from '../../assets/right-arrow.svg';

const Form = () => {
    const dispatch = useDispatch();
    const navigate = useNavigate();
    const [isOtherTopic, setIsOtherTopic] = useState<boolean>(false);
    const [values, setValues] = useState({
        userName: '', surname: '', topic: '', otherTopic: ''
    });
    const [error, setError] = useState(false);

    const handleChange = (e: { target: { name: any; value: string; }; }) => {
        setValues(prev => ({ ...prev, [e.target.name]: e.target.value }));
        if (e.target.value === 'Other') {
            setIsOtherTopic(true);
        };
        if (e.target.value === 'Select' || e.target.value === 'Travel' || e.target.value === 'Cars' || e.target.value === 'Wildlife' || e.target.value === 'Technology') {
            setIsOtherTopic(false);
        };
        setError(false);
    }
    const formSubmit = (e: { preventDefault: () => void; }) => {
        e.preventDefault();
        if (values.userName === '' || values.surname === '' || values.topic === '') {
            setError(true);
            return
        }
        setValues({ userName: '', surname: '', topic: '', otherTopic: '' });
        setIsOtherTopic(false);
        dispatch(addUser({
            userName: values.userName,
            surname: values.surname,
            topic: values.topic,
            otherTopic: values.otherTopic,
        }))
        navigate("/ProfilePic");
    }

    return (
        <div>
            <form className="form" onSubmit={formSubmit} data-testid="form">
                <div className="formControl">
                    <label htmlFor="username">Name<span className='asterisk'>*</span> </label>
                    <input type="text" id="username" className="formInput" value={values.userName} name="userName" onChange={handleChange} />
                </div>
                <div className="formControl">
                    <label htmlFor="surname">Surname<span className='asterisk'>*</span> </label>
                    <input type="text" id="surname" className="formInput" value={values.surname} name="surname" onChange={handleChange} />
                </div>
                <div className="formControl">
                    <label htmlFor="topic">Preferred Topic<span className='asterisk'>*</span> </label>
                    <select id="topic" className="formInput" name="topic" value={values.topic} onChange={handleChange}>
                        <option value="Select">Select</option>
                        <option value="Travel">Travel</option>
                        <option value="Cars">Cars</option>
                        <option value="Wildlife">Wildlife</option>
                        <option value="Technology">Technology</option>
                        <option value="Other">Other</option>
                    </select>
                </div>
                {isOtherTopic &&
                    <div className="formControl">
                        <label htmlFor="otherTopic">Other Topic<span className='asterisk'>*</span></label>
                        <input type="text" id="otherTopic" className="formInput" value={values.otherTopic} name="otherTopic" onChange={handleChange} />
                    </div>
                }
                <button type="submit" className='buttonPrimary' data-testid="formSubmit">  Next &nbsp; <img src={arrow} alt="Arrow" width='20' /></button>
                {error && <p className="error">All Mandatory Fields Are Required.</p>}
            </form>
        </div>
    )
};

export default Form;