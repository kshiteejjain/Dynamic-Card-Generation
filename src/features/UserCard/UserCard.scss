@import '../../app/utils/variable.scss';

h1 {
    text-align: center;
    color: $white;
}

.cardWrapper {
    display: flex;
    align-items: center;
    flex-direction: column;
}

.cardParent {
    display: flex;
    justify-content: center;
    flex-wrap: wrap;
}

.card {
    background-color: $white;
    display: flex;
    flex-direction: column;
    align-items: center;
    width: 21rem;
    position: relative;
    margin: 1rem;
    border-radius: 0.8rem;
    padding-bottom: 3rem;

    &:before {
        background-image: url(../../assets/bg.jpeg);
        content: '';
        position: absolute;
        top: 0;
        width: 100%;
        height: 10rem;
        background-repeat: no-repeat;
        border-radius: 0.8rem 0.8rem 0 0;
    };

    label {
        text-align: left;
        width: 100%;
        padding: 0.5rem 2rem;
        box-sizing: border-box;
        font-weight: 500;
    };
    
    span {
        color: $black;
        font-weight: 300;
    }
}

.cardImage {
    position: relative;
    margin: 4rem 0 2rem 0;
    width: 9rem;
    height: 9rem;
}
