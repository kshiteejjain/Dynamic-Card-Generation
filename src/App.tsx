import React from 'react';
import { Routes, Route} from "react-router-dom";
import './App.scss';
import Form from './features/Form/Form';
import UserCard from './features/UserCard/UserCard';
import ProfilePic from './features/ProfilePic/ProfilePic';
import logo from './assets/logo.svg';

const App = () => {
  return (
    <>
        <header>
        <img data-testid='renderLogo' src={logo} alt="Logo" width='55' />
          <h1 data-testid='renderTitle'>Image Finder</h1>
        </header>

        <div className="container">
            <Routes>
              <Route path="/" element={<Form />} />
              <Route path="*" element={<Form />} />
              <Route path="UserCard" element={<UserCard />} />
              <Route path="ProfilePic" element={<ProfilePic />} />
            </Routes>
        </div>
    </>
  );
}

export default App;
