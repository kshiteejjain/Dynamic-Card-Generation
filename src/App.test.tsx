import React from 'react';
import { renderWithProviders } from '../src/test-utils';
import { screen } from '@testing-library/react'
import App from './App';
import Form from './features/Form/Form';
import UserCard from './features/UserCard/UserCard';
import { createMemoryHistory } from "history";

const mockedUsedNavigate = jest.fn();
jest.mock('react-router-dom', () => ({
  ...jest.requireActual('react-router-dom'),
  useNavigate: () => mockedUsedNavigate,
}));

jest.mock('./app/utils/variable.scss', () => '')

describe('Should Render <App />', () => {
  
  // const renderProvider = () => {
  //   const initialState = {}
  //   const mockStore = configureStore();
  //   let store = mockStore(initialState);
  //   const history = createMemoryHistory({ initialEntries: ['/'] });
  //   render(<Provider store={store} history={history}> <App /> <Form /> <UserCard /> </Provider>)
  // }

  it('renders correctly without error', async () => {
    renderWithProviders(<App />);

    expect(renderWithProviders).toMatchSnapshot();
  });

it('Render Header Logo and Title', () => {
  renderWithProviders(<App />)

  expect(screen.getByTestId('renderLogo')).toBeInTheDocument();
  expect(screen.getByTestId('renderTitle')).toBeInTheDocument();
});

it("Redirects to Form Page", () => {
  const history = createMemoryHistory();
  renderWithProviders(<Form />);

  expect(history.location.pathname).toBe("/");
});

it('Renders Form component by default', async() => {
  renderWithProviders(<Form />);

  const form = screen.getAllByTestId('form');
  await(() => expect(form).toBeInTheDocument())
});

it('Renders UserCard component', async () => {
  renderWithProviders(<UserCard />);

  const form = screen.getAllByTestId('cardWrapper');
  await(() => expect(form).toBeInTheDocument())
});

});