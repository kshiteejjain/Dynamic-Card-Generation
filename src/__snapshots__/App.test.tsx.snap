// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App Component renders correctly without error 1`] = `[Function]`;

exports[`Should Render <App /> renders correctly without error 1`] = `[Function]`;
